File: backend/app/models/supplier.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Supplier(Base):
    __tablename__ = "suppliers"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(180), nullable=False)
    trade_name: Mapped[str | None] = mapped_column(String(180))
    document_number: Mapped[str | None] = mapped_column(String(30), index=True)
    state_registration: Mapped[str | None] = mapped_column(String(40))
    phone: Mapped[str | None] = mapped_column(String(40))
    email: Mapped[str | None] = mapped_column(String(180))
    address_line: Mapped[str | None] = mapped_column(String(180))
    address_number: Mapped[str | None] = mapped_column(String(20))
    address_complement: Mapped[str | None] = mapped_column(String(120))
    neighborhood: Mapped[str | None] = mapped_column(String(120))
    city: Mapped[str | None] = mapped_column(String(120))
    state: Mapped[str | None] = mapped_column(String(2))
    city_code: Mapped[str | None] = mapped_column(String(20))
    zip_code: Mapped[str | None] = mapped_column(String(20))
    notes: Mapped[str | None] = mapped_column(Text)
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())

    stock_entries = relationship("StockEntry", back_populates="supplier")
    product_links = relationship("SupplierProductLink", back_populates="supplier", cascade="all, delete-orphan")

File: admin_app/admin_flutter/lib/models/supplier.dart
class Supplier {
  const Supplier({
    required this.id,
    required this.name,
    this.tradeName,
    this.documentNumber,
    this.stateRegistration,
    this.phone,
    this.email,
    this.addressLine,
    this.city,
    this.state,
    this.notes,
    required this.active,
  });

  final int id;
  final String name;
  final String? tradeName;
  final String? documentNumber;
  final String? stateRegistration;
  final String? phone;
  final String? email;
  final String? addressLine;
  final String? city;
  final String? state;
  final String? notes;
  final bool active;

  factory Supplier.fromJson(Map<String, dynamic> json) {
    return Supplier(
      id: json['id'] as int,
      name: json['name'] as String,
      tradeName: json['trade_name'] as String?,
      documentNumber: json['document_number'] as String?,
      stateRegistration: json['state_registration'] as String?,
      phone: json['phone'] as String?,
      email: json['email'] as String?,
      addressLine: json['address_line'] as String?,
      city: json['city'] as String?,
      state: json['state'] as String?,
      notes: json['notes'] as String?,
      active: json['active'] as bool? ?? true,
    );
  }
}

class SupplierPayload {
  const SupplierPayload({
    required this.name,
    this.tradeName,
    this.documentNumber,
    this.stateRegistration,
    this.phone,
    this.email,
    this.addressLine,
    this.city,
    this.state,
    this.notes,
    this.active = true,
  });

  final String name;
  final String? tradeName;
  final String? documentNumber;
  final String? stateRegistration;
  final String? phone;
  final String? email;
  final String? addressLine;
  final String? city;
  final String? state;
  final String? notes;
  final bool active;

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'trade_name': _emptyToNull(tradeName),
      'document_number': _emptyToNull(documentNumber),
      'state_registration': _emptyToNull(stateRegistration),
      'phone': _emptyToNull(phone),
      'email': _emptyToNull(email),
      'address_line': _emptyToNull(addressLine),
      'city': _emptyToNull(city),
      'state': _emptyToNull(state),
      'notes': _emptyToNull(notes),
      'active': active,
    };
  }
}

String? _emptyToNull(String? value) {
  final trimmed = value?.trim();
  if (trimmed == null || trimmed.isEmpty) return null;
  return trimmed;
}
